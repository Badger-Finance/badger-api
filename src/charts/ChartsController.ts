import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { SettSnapshot } from '../interface/SettSnapshot';
import { SettsService } from '../setts/SettsService';

@Controller('/charts')
export class ChartsController {
  @Inject()
  settsService!: SettsService;

  @Get()
  @ContentType('json')
  getChart(@QueryParams('settName') settName: string, @QueryParams('count') count?: number): Promise<SettSnapshot[]> {
    return this.settsService.getSettSnapshots(settName, count);
  }
}
