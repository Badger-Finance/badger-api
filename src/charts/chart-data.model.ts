import { attribute } from '@aws/dynamodb-data-mapper-annotations';

export abstract class ChartData<T> {
  @attribute()
  id!: string;

  @attribute()
  timestamp!: number;

  abstract toBlankData(): T;
}
