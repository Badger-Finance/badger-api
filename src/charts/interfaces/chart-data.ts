export interface ChartData {
  id: string;
  timestamp: number;
}
