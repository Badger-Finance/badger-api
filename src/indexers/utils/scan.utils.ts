import { Network } from '@badger-dao/sdk';
import { BlocksRangeOptions } from '@badger-dao/sdk/lib/common';

import { getDataMapper } from '../../aws/dynamodb.utils';
import { IndexingMetadata } from '../../aws/models/indexing-metadata.model';
import { MAX_SCAN_RANGE } from '../constants/scan.constants';
import { LastScannedBlockMeta } from '../interfaces/last-scanned-block-meta.interface';

export async function getOrCreateMetadata<T>(task: string): Promise<IndexingMetadata<T>> {
  const mapper = getDataMapper();

  let metadata!: IndexingMetadata<T>;

  try {
    metadata = await mapper.get(Object.assign(new IndexingMetadata<T>(), { task }));
  } catch (e) {
    metadata = Object.assign(new IndexingMetadata<T>(), { task, data: {} });
  }

  return metadata;
}

export function getLastScanedBlock<T extends LastScannedBlockMeta>(
  indexingMetaData: T,
  network: Network,
  address: string,
) {
  return indexingMetaData[network]?.[address]?.lastScannedBlock || 0;
}

export function getScanRangeOpts(
  fromChain: Required<BlocksRangeOptions>,
  lastScannedBlock: number,
): Required<BlocksRangeOptions> {
  const startBlock = lastScannedBlock > 0 ? lastScannedBlock + 1 : fromChain.startBlock;
  let endBlock = fromChain.endBlock;

  if (endBlock - startBlock > MAX_SCAN_RANGE) {
    endBlock = startBlock + MAX_SCAN_RANGE;
  }

  return {
    startBlock,
    endBlock,
  };
}

export async function saveIndexingMetadata<T>(model: IndexingMetadata<T>): Promise<void> {
  const mapper = getDataMapper();

  try {
    await mapper.put(model);
  } catch (e) {
    console.error(`Failed to save indexing metadata for ${model.task}`, e);
  }
}
