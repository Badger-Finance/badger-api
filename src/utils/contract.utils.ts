import { ethers } from 'ethers';
import { RawAbiDefinition } from 'typechain/dist/parser/abiParser';

function getMethodSignature({ inputs, name }: RawAbiDefinition) {
  const params = inputs?.map((input) => input.type).join(',');

  return `${name}(${params})`;
}

function getSignatureHash(signature: string) {
  return ethers.utils.keccak256(ethers.utils.toUtf8Bytes(signature)).toString();
}

function getMethodMeta(method: RawAbiDefinition) {
  const methodSignature = getMethodSignature(method);
  const signatureHash = getSignatureHash(methodSignature);

  return { methodSignature, signatureHash };
}

export function extendAbiMethodsWithMeta(abi: RawAbiDefinition[]) {
  const methods = abi.filter((method) => method.type === 'function');

  return methods.map((method) => ({
    ...getMethodMeta(method),
    ...method,
    decodedParams: [],
  }));
}
