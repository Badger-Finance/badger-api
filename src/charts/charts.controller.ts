import { Controller, Get, Inject, QueryParams, UsePipe } from '@tsed/common';
import { NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { ContentType, Hidden } from '@tsed/schema';
import { ethers } from 'ethers';
import { Chain } from '../chains/config/chain.config';
import { ValidationPipe } from '../common/decorators/validation-pipe';
import { VaultSnapshot } from '../vaults/interfaces/vault-snapshot.interface';
import { ChartsService } from './charts.service';
import { ChartsQueryDto } from './dto/charts-query.dto';
import { ChartGranularity } from './enums/chart-granularity.enum';

@Controller('/charts')
export class ChartsController {
  @Inject()
  chartsService!: ChartsService;

  @Hidden()
  @Get()
  @ContentType('json')
  async getChart(
    @UsePipe(ValidationPipe)
    @QueryParams()
    query: ChartsQueryDto,
  ): Promise<VaultSnapshot[]> {
    const now = new Date();
    const yesterday = new Date();
    yesterday.setHours(now.getHours() - 24);

    const {
      id: settToken,
      chain,
      start = yesterday,
      end = now,
      period = 1,
      granularity = ChartGranularity.HOUR,
    } = query;

    const { isValid, error } = this.isValidGranularityPeriod(start, end, granularity, period);

    if (!isValid && error) {
      throw new UnprocessableEntity(error);
    }

    let checksumContract = settToken;
    try {
      checksumContract = ethers.utils.getAddress(settToken);
    } catch (err) {
      throw new UnprocessableEntity('Invalid contract address');
    }

    const vault = Chain.getChain(chain).vaults.find((vault) => vault.vaultToken === checksumContract);

    if (!vault) {
      throw new NotFound(`${checksumContract} is not a valid sett`);
    }

    return this.chartsService.getChartData(vault, start, end, granularity, period);
  }

  private isValidGranularityPeriod(
    start: Date,
    end: Date,
    granularity: ChartGranularity,
    period: number,
  ): { isValid: boolean; error?: string } {
    const maxDataPointsToReturn = 50;

    if (start >= end) {
      return {
        isValid: false,
        error: '"start" must be strictly less than "end"',
      };
    }

    const returnedDataPoints = ChartsService.getRequestedDataPoints(start, end, granularity, period);
    if (returnedDataPoints > maxDataPointsToReturn) {
      return {
        isValid: false,
        error: 'Too many data points to return. Increase the period or shorten the time range.',
      };
    }

    return {
      isValid: true,
    };
  }
}
