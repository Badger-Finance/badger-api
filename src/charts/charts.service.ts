import { Service } from '@tsed/di';
import { Chain } from '../chains/config/chain.config';
import { SettDefinition } from '../setts/interfaces/sett-definition.interface';
import { SettSnapshot } from '../setts/interfaces/sett-snapshot.interface';
import { getSettSnapshotsInRange } from '../setts/setts.utils';
import { ChartGranularity } from './enums/chart-granularity.enum';

@Service()
export class ChartsService {
  async getChartData(
    chain: Chain,
    sett: SettDefinition,
    start: Date,
    end: Date,
    granularity: ChartGranularity,
    period: number,
  ): Promise<SettSnapshot[]> {
    let snapshots = await getSettSnapshotsInRange(sett, start, end);

    // snapshot granularity @ 30 min intervals, 2 per hour, 48 per day
    const interval = granularity === ChartGranularity.HOUR ? 2 : 48 * period;
    // filter data down to selected granularity period combination
    snapshots = snapshots.filter((_s, i) => i % interval === 0);

    const expectedDataPoints = this.getRequestedDataPoints(start, end, granularity, period);
    if (snapshots.length < expectedDataPoints) {
      let lastSnapshot = snapshots[snapshots.length - 1];
      if (!lastSnapshot) {
        lastSnapshot = {
          address: sett.settToken,
          value: 0,
          balance: 0,
          supply: 0,
          ratio: 1,
          height: await chain.provider.getBlockNumber(),
          timestamp: Date.now(),
        };
      }
      const thirtyMinutesBlocks = parseInt((chain.blocksPerYear / 365 / 24 / 2).toString());
      while (snapshots.length < expectedDataPoints) {
        const newSnapshot: SettSnapshot = {
          address: sett.settToken,
          value: 0,
          balance: 0,
          supply: 0,
          ratio: 1,
          height: lastSnapshot.height - thirtyMinutesBlocks * (interval ?? 1),
          timestamp: lastSnapshot.timestamp - 30 * 60 * 1000 * (interval ?? 1),
        };
        snapshots.push(newSnapshot);
        lastSnapshot = newSnapshot;
      }
    }

    return snapshots;
  }

  getRequestedDataPoints(start: Date, end: Date, granularity: ChartGranularity, period: number): number {
    let returnedDataPoints: number;
    const diffInMs = new Date(end).getTime() - new Date(start).getTime();
    if (granularity === ChartGranularity.DAY) {
      const diffInDays = diffInMs / (1000 * 60 * 60 * 24);
      returnedDataPoints = Math.floor(diffInDays / period);
    } else {
      const diffInHours = diffInMs / (1000 * 60 * 60);
      returnedDataPoints = Math.floor(diffInHours / period);
    }
    return returnedDataPoints;
  }
}
