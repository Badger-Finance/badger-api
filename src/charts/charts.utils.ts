import { ChartTimeFrame, ONE_DAY_MS, ONE_HOUR_MS } from '@badger-dao/sdk';

import { getDataMapper } from '../aws/dynamodb.utils';
import { ChartDataBlob } from '../aws/models/chart-data-blob.model';
import { HistoricVaultSnapshotModel } from '../aws/models/historic-vault-snapshot.model';
import { ChartData } from './chart-data.model';

// list of ChartTimeFrame enums that contain unique capture granularities for searching
export const CHART_GRANULARITY_TIMEFRAMES = [ChartTimeFrame.Max, ChartTimeFrame.Week, ChartTimeFrame.Day];

/**
 *
 * @param id
 * @param timeframe
 * @param data
 * @returns
 */
export function toChartDataBlob<T extends ChartData<T>>(
  id: string,
  timeframe: ChartTimeFrame,
  data: T[]
): ChartDataBlob<T> {
  return Object.assign(new ChartDataBlob(), {
    id,
    timeframe,
    data
  });
}

/**
 *
 * @param namespace
 * @param id
 * @param timeframe
 * @returns
 */
export function toChartDataKey(namespace: string, id: string, timeframe: ChartTimeFrame): string {
  return [namespace, id, timeframe].join('_');
}

/**
 *
 * @param reference
 * @param timestamp
 * @param timeframe
 * @returns
 */
export function shouldUpdate(reference: number, timestamp: number, timeframe: ChartTimeFrame): boolean {
  const difference = reference - timestamp;
  let update = false;
  switch (timeframe) {
    case ChartTimeFrame.Max:
    case ChartTimeFrame.Year:
    case ChartTimeFrame.ThreeMonth:
    case ChartTimeFrame.Month:
    case ChartTimeFrame.YTD:
      update = difference >= ONE_DAY_MS;
      break;
    case ChartTimeFrame.Week:
      update = difference >= ONE_HOUR_MS * 6;
      break;
    default:
      update = difference >= ONE_HOUR_MS;
  }
  return update;
}

/**
 *
 * @param reference
 * @param timestamp
 * @param timeframe
 * @returns
 */
export function shouldTrim(reference: number, timestamp: number, timeframe: ChartTimeFrame): boolean {
  const difference = reference - timestamp;
  let update;
  switch (timeframe) {
    case ChartTimeFrame.Year:
      update = difference >= ONE_DAY_MS * 365;
      break;
    case ChartTimeFrame.ThreeMonth:
      update = difference >= ONE_DAY_MS * 90;
      break;
    case ChartTimeFrame.Month:
      update = difference >= ONE_DAY_MS * 30;
      break;
    case ChartTimeFrame.Week:
      update = difference >= ONE_DAY_MS * 7;
      break;
    case ChartTimeFrame.Day:
      update = difference >= ONE_DAY_MS;
      break;
    // Year to date needs a full clear - so we will never trim
    case ChartTimeFrame.YTD:
    default:
      update = false;
  }
  return update;
}

/**
 *
 * @param namespace
 * @param snapshot
 */
export async function updateSnapshots<T extends ChartData<T>>(namespace: string, snapshot: T) {
  const mapper = getDataMapper();
  const isFirstOfYear = (date: Date) => date.getDay() === 0 && date.getMonth() === 0;

  const { id, timestamp: now } = snapshot;
  for (const timeframe of Object.values(ChartTimeFrame)) {
    const searchKey = Object.assign(new ChartDataBlob<T>(), {
      id: toChartDataKey(namespace, id, timeframe)
    });

    let cachedChart: ChartDataBlob<T> | undefined;
    try {
      cachedChart = await mapper.get(searchKey);
      if (timeframe === ChartTimeFrame.YTD) {
        const date = new Date(now);
        if (isFirstOfYear(date) && cachedChart.data.length > 1) {
          // new year, force a new object to be created
          cachedChart = undefined;
        }
      }
    } catch (err) {
      console.debug({ message: 'Unable to query cached chart, may simply not exist', err });
    } // no item found

    let updateCache = false;
    if (!cachedChart) {
      const blob = toChartDataBlob(searchKey.id, timeframe, []);
      console.debug(`Create blob for ${searchKey.id}`);
      try {
        cachedChart = await mapper.put(blob);
      } catch (err) {
        console.error({ message: 'Unable to save blob', err });
      }
      updateCache = true;
    } else {
      const { timeframe, data } = cachedChart;
      const recentSnapshot = data[0].timestamp;
      updateCache = shouldUpdate(now, recentSnapshot, timeframe);

      const lastSnapshot = data[data.length - 1].timestamp;
      if (shouldTrim(now, lastSnapshot, timeframe)) {
        cachedChart.data = cachedChart.data.slice(0, data.length - 1);
      }
    }

    if (updateCache && cachedChart) {
      cachedChart.data.splice(0, 0, snapshot);
      console.debug(`Update ${searchKey.id} (${cachedChart.data.length} entries)`);
      try {
        await mapper.put(cachedChart);
      } catch (err) {
        console.error({ message: 'Unable to save blob', err });
      }
    }
  }
}

export async function queryVaultCharts(id: string): Promise<HistoricVaultSnapshotModel[]> {
  try {
    const mapper = getDataMapper();
    for await (const item of mapper.query(ChartDataBlob, { id }, { limit: 1, scanIndexForward: false })) {
      return item.data as HistoricVaultSnapshotModel[];
    }
    return [];
  } catch (err) {
    console.error(err);
    return [];
  }
}
