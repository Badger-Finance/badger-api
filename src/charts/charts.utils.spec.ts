import { ChartTimeFrame } from '@badger-dao/sdk';
import { ChartData } from './chart-data.model';
import { toChartDataBlob } from './charts.utils';

describe('charts.utils', () => {
  describe('toChartDataBlob', () => {
    it('converts arbitrary data array into a chart data blob', () => {
      class AnimalData extends ChartData<AnimalData> {
        animal!: string;
        noise!: string;

        toBlankData(): AnimalData {
          const copy = JSON.parse(JSON.stringify(this));
          copy.animal = '';
          copy.noise = '';
          return copy;
        }
      }

      const arbitraryData: AnimalData[] = [
        Object.assign(new AnimalData(), {
          id: 'sighting-0',
          timestamp: 10,
          animal: 'cow',
          noise: 'moo',
        }),
        Object.assign(new AnimalData(), {
          id: 'sighting-1',
          timestamp: 20,
          animal: 'dog',
          noise: 'bark',
        }),
        Object.assign(new AnimalData(), {
          id: 'sighting-2',
          timestamp: 30,
          animal: 'cat',
          noise: 'lmeow',
        }),
      ];
      const result = toChartDataBlob('animals', ChartTimeFrame.Week, arbitraryData);
      expect(result).toMatchSnapshot();
    });
  });

  describe('', () => {});
});
