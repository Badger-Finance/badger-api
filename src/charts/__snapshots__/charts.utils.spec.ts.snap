// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`charts.utils toChartDataBlob converts arbitrary data array into a chart data blob 1`] = `
ChartDataBlob {
  "data": Array [
    AnimalData {
      "animal": "cow",
      "id": "sighting-0",
      "noise": "moo",
      "timestamp": 10,
    },
    AnimalData {
      "animal": "dog",
      "id": "sighting-1",
      "noise": "bark",
      "timestamp": 20,
    },
    AnimalData {
      "animal": "cat",
      "id": "sighting-2",
      "noise": "lmeow",
      "timestamp": 30,
    },
  ],
  "id": "animals",
  "timeframe": "7d",
}
`;
